peps/pep-0788.rst

```diff
@@ -0,0 +1,603 @@
+  crashing it with a segmentation fault, preventing further execution.
+- When they're called before finalization, they force the thread to be
+  "daemon," meaning that the interpreter won't wait for it to reach any point
+  of execution. This is mostly frustrating for developers, but can lead to
+  deadlocks!
+- Subinterpreters don't play nicely with them, because they all assume that
+  the main interpreter is the only one that exists. A fresh thread (*i.e.*,
+  has never had a thread state) that calls :c:func:`PyGILState_Ensure` will
+  always be for the main interpreter.
+- The term "GIL" in the name is quite confusing for users of free-threaded
+  Python. There isn't a GIL, why do they still have to call it?
+
+This PEP intends to fix all of these issues by providing
+:c:func:`PyThreadState_Ensure` and :c:func:`PyThreadState_Release` as a more
+correct and safer replacement for :c:func:`PyGILState_Ensure` and
+:c:func:`PyGILState_Release`. For example:
+
+.. code-block:: C
+
+        if (PyThreadState_Ensure(interp) < 0) {
+            fputs("Python is shutting down", stderr);
+            return;
+        }
+
+        /* Interact with Python, without worrying about finalization. */
+        // ...
+
+        PyThreadState_Release();
+
+Motivation
+==========
+
+Native threads will always hang during finalization
+---------------------------------------------------
+
+Many codebases might need to call Python code in highly-asynchronous
+situations where the interpreter is already finalizing, or might finalize, and want to continue
+running code after the Python call. This desire has been
+`brought up by users <https://discuss.python.org/t/safely-using-the-c-api-when-python-might-shut-down/78850/>`_.
+For example, a callback that wants to call Python code might be invoked by:
+
+- A kernel has finished running on the GPU
+- A network packet was received.
+- A thread has quit, and the C++ library is executing static finalizers of
+  thread local storage.
+
+In the current C API, any non-Python thread (*i.e.*, not created by
+:mod:`threading`) is considered to be "daemon," meaning that the interpreter
+won't wait on that thread to finalize. Instead, the interpreter will hang the
+thread when it goes to `attach <attached thread state>`_ a `thread state`_,
+making it unusable past that point. Attaching a thread state can happen at
+any point when invoking Python, such as releasing the GIL in-between bytecode
+instructions, or when a C function exits a :c:macro:`Py_BEGIN_ALLOW_THREADS`
+block.
+
+This means that any non-Python thread may be terminated at any point, which
+is severely limiting for users who want to do more than just execute Python
+code in their stream of calls (for example, C++ executing finalizers in
+*addition* to calling Python).
+
+``Py_IsFinalizing`` is insufficient
+***********************************
+
+The `docs <https://docs.python.org/3/c-api/init.html#c.PyGILState_Ensure>`_
```

Comment:
'The docs currently recommend' feels a bit too vague. Perhaps "The best current solution to the finalisation problem is implementing checks using Py_IsFinalizing, as recommended in the C API documentation. Unfortunately..."

```suggestion
The :c:func:`docs <PyGILState_Ensure>`
```